web-admin/src/features/projects/github/GithubData.ts

```diff
@@ -0,0 +1,144 @@
+          queryClient,
+        },
+      }).subscribe(set),
+  ) as ReturnType<
+    typeof createAdminServiceListGithubUserRepos<
+      Awaited<ReturnType<typeof adminServiceListGithubUserRepos>>,
+      RpcStatus
+    >
+  >;
+
+  private userPromptWindow: Window | undefined;
+  private windowCheckTimer: ReturnType<typeof setInterval>;
+
+  public readonly status = derived(
+    [this.userStatus, this.userRepos],
+    ([userStatus, userRepos]) => {
+      if (userStatus.isFetching || userRepos.isFetching) {
+        return {
+          isFetching: true,
+          error: undefined,
+        };
+      }
+
+      return {
+        isFetching: false,
+        error: userStatus.error ?? userRepos.error,
+      };
+    },
+  );
+
+  /**
+   * Marks the repo selection dialog to be opened.
+   * If user doesn't have access, opens grant access page.
+   */
+  public async startRepoSelection() {
+    this.repoSelectionOpen.set(true);
+
+    await waitUntil(() => !get(this.userStatus).isFetching);
+    const userStatus = get(this.userStatus).data;
+    if (!userStatus || userStatus?.hasAccess) {
+      return;
+    }
+
+    this.openGithubConnectWindow(userStatus.grantAccessUrl);
+  }
+
+  /**
+   * Used to reselect connected repos.
+   * Opens the grantAccessUrl page to achieve this.
+   */
+  public async reselectRepos() {
+    await waitUntil(() => !get(this.userStatus).isFetching);
+    const userStatus = get(this.userStatus).data;
+    if (!userStatus?.grantAccessUrl) {
+      return;
+    }
+
+    this.openGithubConnectWindow(userStatus.grantAccessUrl);
+  }
+
+  public async refetch() {
+    const userStatus = get(this.userStatus).data;
+    if (!userStatus?.hasAccess) {
+      // refetch status if had no access
+      await queryClient.refetchQueries(
+        getAdminServiceGetGithubUserStatusQueryKey(),
+      );
+
+      await waitUntil(() => !get(this.userStatus).isFetching);
+      if (!get(this.userStatus).data?.hasAccess) {
+        this.githubConnectionFailed.set(true);
+      } else {
+        this.githubConnectionFailed.set(false);
+      }
+    } else {
+      this.githubConnectionFailed.set(false);
+
+      // else refetch the list of repos
+      await queryClient.refetchQueries(
+        getAdminServiceListGithubUserReposQueryKey(),
+      );
+    }
+  }
+
+  private openGithubConnectWindow(url: string) {
+    try {
+      // safeguard try catch
+      this.userPromptWindow?.close();
+    } catch (e) {
+      // no-op
+    }
+    if (this.windowCheckTimer) clearInterval(this.windowCheckTimer);
+    this.userPromptWindow = window.open(url, "", "width=1024,height=600");
+
+    // periodically check if the new window was closed
+    this.windowCheckTimer = setInterval(() => {
+      if (!this.userPromptWindow.closed) return;
+      clearInterval(this.windowCheckTimer);
+      this.windowCheckTimer = undefined;
+      this.userPromptWindow = undefined;
+      void this.refetch();
```

Comment:
This looks like it's refetching when the `userPromptWindow` is _open_. Shouldn't refetch only once the `userPromptWindow` is _closed_?